src/browser/AxePage.ts

```diff
@@ -86,82 +106,20 @@ export function getRunOptions(
- * https://github.com/dequelabs/axe-core/issues/1041).
+ * Used to prevent concurrent runs of `axe.run`, which breaks (see https://github.com/dequelabs/axe-core/issues/1041).
+ * This should never happen, but in the past errors during rendering at the right/wrong time has
+ * caused the next test to start before the previous has stopped.
  *
- * We avoid that by forcing one `axe.run` call to finish before the next one can start. Got the
- * idea from https://github.com/dequelabs/agnostic-axe/pull/6.
+ * Got the idea from https://github.com/dequelabs/agnostic-axe/pull/6.
  */
 function addPromiseQueue() {
-  type QueuedPromise<T> = {
-    promiseCreator: () => Promise<T>;
-    resolve: (value: T | PromiseLike<T>) => void;
-    reject: (reason?: unknown) => void;
-  };
-
-  /**
-   * Queue of promises, which forces any promises added to it to run one at a time.
-   *
-   * This was originally implemented as an ES6 class. But I ran into a lot of issues with how the
-   * class's properties were compiled not working in different environments and browsers, and even
-   * breaking when Babel was updated.
-   *
-   * To avoid that, I've instead implemented this as a function that maintains some state within a
-   * closure. Since there are no class properties (which can be compiled by Babel in different ways),
-   * there are no more problems.
-   */
-  function PromiseQueue<T>() {
-    const pending: QueuedPromise<T>[] = [];
-    let working = false;
-
-    /**
-     * Add a promise to the queue. Returns a promise that is resolved or rejected when the added
-     * promise eventually resolves or rejects.
-     */
-    function add(promiseCreator: () => Promise<T>): Promise<T> {
-      return new Promise((resolve, reject) => {
-        pending.push({promiseCreator, resolve, reject});
-        dequeue();
-      });
-    }
-
-    /**
-     * Run the next promise in the queue.
-     */
-    function dequeue() {
-      // If we're already working on a promise, do nothing.
-      if (working) {
-        return;
-      }
-
-      const nextPromise = pending.shift();
+  let queue = Promise.resolve();
 
-      // If there are no promises to work on, do nothing.
-      if (!nextPromise) {
-        return;
-      }
-
-      working = true;
-
-      // Execute the promise. When it's done, start working on the next.
-      nextPromise
-        .promiseCreator()
-        .then(nextPromise.resolve, nextPromise.reject)
-        .finally(() => {
-          working = false;
-          dequeue();
-        });
-    }
-
-    return {add};
-  }
-
-  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
-  // @ts-ignore This function executes in a browser context.
-  window.axeQueue = PromiseQueue<AxeResults>();
+  window.enqueuePromise = function <T>(createPromise: () => Promise<T>) {
+    return new Promise<T>((resolve, reject) => {
+      queue = queue.then(createPromise).then(resolve).catch(reject);
```

Comment:
Am I understanding correctly that the first `.then` invokes the promise at the end of the queue, and then `createPromise` is invoked, adding on to the queue? 

Does the second `.then` then resolve the one just created by `createPromise`?

Sorry a lot of promises going around 😅

Reply:
Sort of!

`.then` creates a new promise off of the current `queue` (which of course is then set as the current `queue`). It's resolved with the value the created promise is resolved with...

Might be more clear if I wrote it out as

```js
queue = queue.then(() => {
  return createPromise();
})
```

of if we made it less generic and inlined the promise creation (for illustration purposes), it would look something like

```js
// For each story
queue = queue.then(() => {
  return axe.run(...);
})
```

Going even deeper, if we unrolled the loop it would be like

```js
Promise.resolve()
  .then(() => axe.run(...)) // Story 1
  .then(() => axe.run(...)) // Story 2
  .then(() => axe.run(...)) // Story 3
  // etc.
```

Hope that helps!

Reply:
I guess the other way to say it is that `.then` creates a promise, and `createPromise` creates a different promise, but they are linked